File: engine/host/browser_process/browser_process.c
/* THE BROWSER PROCESS — one WASM instance per EXTENSION, beside the per-document RENDERER instances.
 *
 * WHAT IT IS. `engine/host/main.c` is the renderer entry: one instance per origin-keyed agent cluster, running
 * an untrusted bundle over a real DOM (SECURITY.md). This is the OTHER entry — the trusted counterpart that
 * runs the algorithms a real browser keeps OUT of the renderer, and it holds no document, no realm, no page
 * script and no frontier. It is what a Chromium reader means by the browser process: the network service's
 * algorithms live here, the renderer is TOLD their results, and CORB is the reason the direction matters.
 *
 * WHO CREATES IT AND WHO OWNS ITS LIFETIME. The OFFSCREEN document, once, when it starts — the same zone that
 * provisions a renderer instance per document and routes between them (SECURITY.md: "the engine NAMES its own
 * child documents; the offscreen ROUTES them"). There is exactly ONE, because it is keyed on nothing: it holds
 * no per-document state to be keyed by, and a second one would be a second answer to a question with one
 * answer. It dies with the offscreen document, which is also when every renderer instance dies, so it never
 * outlives or is outlived by the instances that ask it. Nothing else may hold its Module handle: a renderer is
 * a separate WASM instance and has no way to call these entries at all — its requests reach the zone as
 * `qjs_host_requests` lines and the zone is what relays them here.
 *
 * AND TODAY THE ZONE THAT PROVISIONS ONE IS `engine/browser_process.mjs`, WHICH IS A STATEMENT ABOUT THIS TREE
 * AND NOT ABOUT THE DESIGN. That driver plays the offscreen — provisions the instance, asks as the zone, and
 * relays a renderer's request as the zone would — for the same reason `engine/route.mjs` is the only thing that
 * provisions a second renderer: a transport no host has driven is a design that has never run (SECURITY.md).
 * `extension/offscreen-brain.js` does not create one yet, and it will on the day it has something to do with
 * the answer — which is the reply record's computed type, read by `solver/reply_decode.c` and written by
 * nobody. A `createBrowserProcess()` in the offscreen ahead of that consumer would be an instance nothing asks.
 *
 * WHY THE ASKER IS AN ARGUMENT AND NOT A FIELD OF THE RECORD. `requester` is stated by the TRUSTED ZONE on
 * every call, exactly as `qjs_route`'s `sender_origin` is and for the identical reason: an identity the
 * untrusted side can write into the text it sends is not an identity. NULL is the zone speaking for itself —
 * a positive statement, not a hole a caller left — and a non-NULL value is the document id of the renderer
 * instance whose request line the zone is relaying. The zone knows which instance handed it a line; the line
 * cannot say.
 *
 * WHO MAY ASK WHAT IS DECLARED BESIDE THE OPERATION, never asked at the call — the same rule §7.2.5.1's
 * cross-origin allowlist follows, and for the same reason: a check written at the entry is one check for a
 * growing table, and the first operation added without a thought about the asker inherits whatever that one
 * check happened to say. Today the table has one row and it is zone-only, so a renderer may ask for NOTHING,
 * and the answer to "what is a renderer allowed to ask for" is a column in the table rather than a sentence
 * here that will go stale.
 *
 * WHAT CROSSES IS TEXT AND IT CARRIES ITS TYPE (SECURITY.md). A request is one TAB-separated record —
 * `<operation><TAB><operands…>` — which is the grammar `qjs_host_requests` and `qjs_perform` already speak, so
 * a relayed renderer request needs no re-encoding on the way through the zone. Bytes travel BESIDE the record
 * and never inside it, exactly as `qjs_provide` and `qjs_host_answer` carry a response body: §7's input is a
 * byte sequence and every way of putting one in text is an encode or a decode performed by a zone that has no
 * business doing either. The ANSWER is the operation's own; `mime.sniff` answers with WHATWG MIME Sniffing
 * §4.5's serialization of a MIME type, which is the standard's own string form of the record and parses back
 * through §4.4 on the far side — one grammar, owned by one component, on both sides of the boundary.
 *
 * NO REALM, AND THAT IS NOT THE SAME QUESTION AS THE ONE SECURITY.md ANSWERS ELSEWHERE. "A PEER ANSWERS BY
 * RUNNING A PROGRAM, never by reading a property from C" is about an AGENT answering a cross-document read:
 * the read is an IDL getter, it belongs to a document's timelines, and it may have to suspend. This process
 * holds no document and no timeline — §7 is a pure function of a header value and some bytes — so an answer
 * here is a return value and the entry is synchronous. The day an operation here has to WAIT (a real network
 * fetch), it gains the ask-then-answer split the renderer seam already has; inventing that split now for an
 * algorithm that cannot use it would be scaffolding with nothing behind it.
 */
#include <stdbool.h>
#include <stdlib.h>
#include <string.h>

#include "check.h"
#include "browser_process/network/mime_sniff.h"
#include "core/mime/mime_type.h"

#ifdef __EMSCRIPTEN__
#include <emscripten.h>
#define BP_EXPORT EMSCRIPTEN_KEEPALIVE
#else
#define BP_EXPORT
#endif

/* ── the operations ──────────────────────────────────────────────────────────────────────────────────────── */

/* WHATWG MIME Sniffing §7 — the COMPUTED MIME type of a response.
 *
 * `args` is the record's remainder: `<no-sniff><TAB><Content-Type value>`, and the Content-Type field is LAST
 * because it is the remainder — a header value may contain HTAB, and the §7.4-notice grammar puts a policy
 * container last for the same reason.
 *
 * A RESPONSE WITH NO `Content-Type` HEADER AT ALL AND ONE WITH AN EMPTY VALUE ARE DIFFERENT FACTS, and the
 * record says which by its FIELD COUNT: two fields is Fetch's "get a header" returning null, three fields is a
 * header whose value is the third. §5.1 happens to reach the same "supplied type is undefined" for both, and
 * that is exactly why the distinction has to live here rather than in whichever caller first noticed it did
 * not matter — a record that cannot say a header is absent has already lost the fact.
 *
 * `no_sniff` is §5's no-sniff flag, whose one source is `X-Content-Type-Options: nosniff`. The ZONE reads that
 * header and states the answer, because it is the zone that holds the response's header list; this process
 * never reaches back for a second fact about a response it was handed one fact about. */
static char *op_mime_sniff(const char *args, const unsigned char *header, size_t header_n)
{
    MimeType computed;
    const char *content_type_value;
    char *serialized;

    /* THE GRAMMAR IS CHECKED, NOT DCHECKED, and the two run in this order for a reason beyond diagnosis: the
       first establishes that `args[0]` is not the terminator, which is what makes reading `args[1]` a read
       inside the string. A dev-only assertion would leave the release build parsing a record it cannot parse —
       an out-of-bounds byte, then a Content-Type field silently read as absent — and there is no answer to give
       for a record this process does not understand while the caller is about to read one. */
    CHECK(args[0] == '0' || args[0] == '1',
          "a `mime.sniff` record's no-sniff field is not a flag — it is §5's no-sniff flag and crosses as `0` "
          "or `1` so that it carries its type, and a zone that wrote anything else stated no flag at all");
    CHECK(args[1] == '\0' || args[1] == '\t',
          "a `mime.sniff` record's no-sniff field ran on past its flag — the field is one byte and the next "
          "byte is either the end of the record or the separator before the Content-Type value, so anything "
          "else means the operands were assembled in an order this operation does not read");

    /* §5.1's supplied-type input: NULL for "the response carries no Content-Type header". */
    content_type_value = (args[1] == '\t') ? args + 2 : NULL;

    mime_sniff_compute(&computed, content_type_value, args[0] == '1', header, header_n);
    /* §4.5's SERIALIZATION IS THE ANSWER, and §7 has no failure outcome — every path through it ends in a MIME
       type, down to `application/octet-stream` — so there is no empty answer for a caller to read as absence. */
    serialized = mime_type_serialize(&computed);
    CHECK(serialized != NULL, "browser process: OOM serializing §7's computed MIME type");
    mime_type_free(&computed);
    return serialized;
}

/* WHO MAY ASK, BESIDE WHAT THEY ASK FOR. `renderer_may_ask` is the whole authorization surface of this process:
   an operation declares it at its own row, and a row added without one does not compile. It reads `false`
   everywhere today because a renderer has nothing it is entitled to ask this process for yet — a renderer's
   business with the browser process is network and storage, and neither of those operations exists here. */
typedef struct {
    const char *op;
    bool        renderer_may_ask;
    char      *(*run)(const char *args, const unsigned char *header, size_t header_n);
} BpOperation;

static const BpOperation OPERATIONS[] = {
    { "mime.sniff", false, op_mime_sniff },
};

/* ── the ABI ─────────────────────────────────────────────────────────────────────────────────────────────── */

/* THE ANSWER LIVES UNTIL THE NEXT REQUEST, which is the contract `qjs_pending` and `qjs_result` already have:
   the zone reads the string out of linear memory before it asks anything else. Freed at the TOP of the next
   call rather than at the bottom of this one, so the pointer this entry returns is never a pointer into freed
   memory. */
static char *g_answer;

BP_EXPORT const char *bp_perform(const char *requester, const char *record,
                                 const char *header, unsigned header_n)
{
    const char *tab;
    size_t op_n, i;

    DCHECK(requester == NULL || *requester,
           "a request arrived from a renderer with an empty name — NULL is how the zone says it is asking for "
           "itself, so an empty string is a name the zone failed to state rather than an absence it stated");
    DCHECK(record != NULL && *record, "the browser process was asked to perform nothing at all");
    DCHECK(header != NULL || header_n == 0,
           "a request's bytes arrived as a null pointer with a length — the pointer and the length describe ONE "
           "byte sequence, so a request carrying no bytes passes both as nothing");

    free(g_answer);
    g_answer = NULL;

    tab = strchr(record, '\t');
    /* CHECKED FOR THE SAME REASON THE FIELD SHAPE ABOVE IS: with no separator there is no operation name to
       find and no operands to run it on, and the pointer arithmetic below would be performed on a null. Every
       operation this process has takes operands, so a record without a TAB is one whose operands were dropped
       somewhere on the way — a caller and a build shipped out of step, not a state to continue past. */
    CHECK(tab != NULL,
          "a request record carries an operation and no operands — the two are separated by a TAB, so a record "
          "without one is either an operation name that lost its arguments or a caller writing a grammar this "
          "process does not speak");
    op_n = (size_t)(tab - record);

    for (i = 0; i < sizeof OPERATIONS / sizeof *OPERATIONS; i++) {
        if (strlen(OPERATIONS[i].op) != op_n || memcmp(record, OPERATIONS[i].op, op_n) != 0)
            continue;
        /* THE SECURITY BOUNDARY, AND IT IS FATAL IN RELEASE TOO. A renderer reaching an operation it is not
           entitled to is the untrusted side deciding what it may be told, which is the one thing this process
           exists to stop — so it is a CHECK and not a DCHECK, and proceeding past it in a shipped build would
           be worse than crashing. */
        CHECK(requester == NULL || OPERATIONS[i].renderer_may_ask,
              "a RENDERER asked the browser process for an operation declared zone-only — the renderer runs the "
              "untrusted bundle, and an answer it is not entitled to is surface the page could never have "
              "obtained, reported as a fact about the page");
        g_answer = OPERATIONS[i].run(tab + 1, (const unsigned char *)header, (size_t)header_n);
        DCHECK(g_answer != NULL,
               "an operation returned no answer — every operation here answers, because a caller that cannot "
               "tell an absent answer from an empty one will read the next request's answer as this one's");
        return g_answer;
    }
    /* AN OPERATION THIS BUILD DOES NOT HAVE IS FATAL IN RELEASE TOO, because there is no answer to give and the
       caller is about to read one: the extension ships this wasm and the zone that drives it together, so a
       name they disagree on is not a state to continue past. */
    CHECK_FAIL("the browser process was asked for an operation it does not have — the `OPERATIONS` table above "
               "is the whole of what this process answers, and a zone naming something else is a caller and a "
               "build that shipped out of step with each other");
}

File: engine/browser_process.mjs
/* THE BROWSER PROCESS'S DRIVER —  node engine/browser_process.mjs  (after node engine/build.mjs).
 *
 * IT IS IN THE REPOSITORY FOR THE REASON engine/route.mjs IS, one process over. SECURITY.md: a host that cannot
 * provision a second instance has not tested the transport, and every cross-instance mechanism is then a design
 * that has never run. route.mjs provisions a second RENDERER; this provisions the BROWSER PROCESS — the trusted
 * instance beside them — and plays the offscreen, which is the only zone that may hold its Module handle.
 *
 * WHAT IT ASKS. One operation: WHATWG MIME Sniffing §7, the computed MIME type of a response. That algorithm
 * was written into `engine/host/browser/core/mime/mime_sniff.c` and called from `solver/reply_decode.c` —
 * inside the RENDERER, which is the process a real browser never lets compute it, because CORB gates on the
 * result and a renderer that classifies a body for itself can mine one it would have been handed as an opaque
 * empty response. The implementation was correct; it was HOUSED wrongly, and it now lives in this process and
 * links into this artifact alone.
 *
 * THE CASES ARE §7'S OWN BRANCHES, not a sample of them: a resource with no Content-Type at all, the
 * `unknown/unknown` essence step 2 names, §7 step 1's XML-or-HTML shortcut, step 3's no-sniff flag, step 4's
 * Apache bug, step 6's image re-match, and the supplied type §7 leaves alone that this engine's one former
 * caller keyed on. Each states what the standard answers and why, so a disagreement names the step.
 *
 * AND IT DRIVES THE AUTHORIZATION BOUNDARY, which is the half a driver usually leaves as a claim. `mime.sniff`
 * is declared zone-only beside the operation, so a renderer asking for it must abort — a CHECK, fatal in dev
 * and in release, because a renderer being told what it is not entitled to is the one thing this process
 * exists to stop. A boundary nothing has ever crossed is a boundary nobody has measured, so this file
 * re-executes ITSELF with `--as-renderer`, in a child process, and requires that child to die: the parent
 * asserts the non-zero exit AND the `@E` line naming the operation. Two spawns for the same reason
 * `build.mjs`'s cold round trip is two — an abort cannot be observed from inside the process it kills. */
import { dirname, join } from 'node:path';
import { fileURLToPath } from 'node:url';
import { spawnSync } from 'node:child_process';

const ENGINE = dirname(fileURLToPath(import.meta.url));
const AS_RENDERER = process.argv.includes('--as-renderer');

const factory = await import(join(ENGINE, '..', 'extension', 'lib', 'qjs', 'browser_process.mjs'));
const boot = factory.default ?? factory;

/* THE OFFSCREEN'S ONE INSTANCE. It is created here and never again: this process is keyed on nothing, holds no
   per-document state, and a second one would be a second answer to a question that has one. */
const M = await boot();

const cs = (s) => {
  if (s === null) return 0;   /* the ZONE speaking for itself, and Fetch's "get a header" returning null */
  const n = M.lengthBytesUTF8(s) + 1, p = M._malloc(n);
  M.stringToUTF8(s, p, n);
  return p;
};

/* THE BYTES TRAVEL BESIDE THE RECORD, never inside it — §5.2's resource header is a BYTE SEQUENCE, and every
   way of putting one in text is an encode or a decode performed by a zone with no business doing either. A
   fixture written as source text is ENCODED here, which is a fixture's own business and never a decode. */
const bytes = (b) => {
  const u8 = typeof b === 'string' ? new Uint8Array([...b].map((c) => c.charCodeAt(0))) : b;
  const p = M._malloc(u8.length + 1);
  M.HEAPU8.set(u8, p);
  return [p, u8.length];
};

/* ONE REQUEST, AS THE ZONE MAKES IT. `requester` is NULL for the zone itself and a renderer's document id when
   the zone is RELAYING that renderer's request — stated here and never read out of the record, exactly as
   `qjs_route`'s sender origin is stated by the zone and never minted by an engine. */
function perform(requester, record, body) {
  const [p, n] = bytes(body);
  try {
    return String(M.ccall('bp_perform', 'string', ['number', 'number', 'number', 'number'],
                          [cs(requester), cs(record), p, n]) ?? '');
  } finally { M._free(p); }
}

/* THE RECORD FOR §7. Two fields is "this response carries no Content-Type header" (Fetch's "get a header"
   returning null); three is a header whose VALUE is the third field, empty value included. The value is last
   because it is the remainder — a header value may contain HTAB. */
const sniff = (contentType, noSniff, body) =>
  perform(null, 'mime.sniff\t' + (noSniff ? '1' : '0') + (contentType === null ? '' : '\t' + contentType), body);

/* ── the renderer's half: a child process, whose job is to die ─────────────────────────────────────────── */
if (AS_RENDERER) {
  /* `d1` is a renderer instance's document id — the name route.mjs's root document carries. The zone would be
     relaying this instance's request; the operation is declared zone-only, so the CHECK fires and this process
     never reaches the line below. */
  const answer = perform('d1', 'mime.sniff\t0\ttext/html', '<!doctype html>');
  console.error('[browser-process] the zone-only boundary DID NOT HOLD: a renderer was answered ' + answer);
  process.exit(0);   /* a zero exit is what the parent reads as the boundary having failed */
}

/* ── §7, case by case ──────────────────────────────────────────────────────────────────────────────────── */

/* §6.1's own byte patterns, built from the byte values the table writes rather than from source-text escapes.
   PNG's signature carries 0x1A, which is also a §3 BINARY DATA BYTE — which is what makes it the right body for
   §7.2's half of the Apache-bug case below, and why the two cases share it. */
const octets = (...b) => String.fromCharCode(...b);
const GIF = 'GIF89a' + octets(0x20, 0x20);
const PNG = octets(0x89) + 'PNG' + octets(0x0D, 0x0A, 0x1A, 0x0A);

const CASES = [
  /* §7 step 2 → §7.1: no supplied type at all, so the answer comes from §6.1's image table. */
  [null, false, GIF, 'image/gif',
   '§7 step 2 with an undefined supplied type runs §7.1, whose image table matches GIF89a'],
  /* §7 step 2 again — `unknown/unknown` is one of the three essences the step names, so a server that states
     it is stating nothing and the bytes decide. */
  ['unknown/unknown', false, PNG, 'image/png',
   "§7 step 2 treats `unknown/unknown` as no statement at all, so §7.1's image table answers"],
  /* §7 step 1: an HTML supplied type is returned BEFORE any byte is looked at — the step that keeps sniffing
     from ever upgrading a resource INTO a scriptable type it was not declared as. The parameter rides along,
     because §4.5 serializes the record and the record has one. */
  ['text/html;charset=utf-8', false, GIF, 'text/html;charset=utf-8',
   '§7 step 1 returns an HTML supplied type unchanged, bytes unread, parameters intact'],
  /* §7 step 3: the no-sniff flag stops the algorithm at the supplied type. The same bytes as the case above
     it, so the flag is the only thing that differs and the answer changes because of it. */
  ['text/plain;charset=utf-8', true, GIF, 'text/plain;charset=utf-8',
   "§7 step 3 returns the supplied type when the no-sniff flag is set — this is `X-Content-Type-Options: " +
   'nosniff`, and the identical body sniffs to image/gif without it'],
  /* §7 step 4 → §7.2: `text/plain` EXACTLY is §5.1's check-for-apache-bug, and a body with a binary data byte
     is then §7.2's `application/octet-stream`. */
  ['text/plain', false, PNG, 'application/octet-stream',
   "§7 step 4's Apache-bug value sends a binary body to §7.2, which answers application/octet-stream"],
  /* And the same value with a body §7.2 finds no binary data byte in. */
  ['text/plain', false, 'plain enough', 'text/plain',
   '§7.2 answers text/plain for a body with no binary data byte'],
  /* §7 step 6: an IMAGE supplied type is re-matched against §6.1's table, and a server that mislabelled a PNG
     as a GIF is corrected by its bytes. */
  ['image/gif', false, PNG, 'image/png',
   "§7 step 6 re-matches an image supplied type against §6.1's patterns, so mislabelled bytes win"],
  /* THE CASE THIS ENGINE'S FORMER CALLER KEYED ON. `solver/reply_decode.c` reads a React Flight stream, and
     `text/x-component` is touched by no rule in §7 — which is the measurement its own comment cites for saying
     the sniff decided zero of the endpoints in the @H surface. It has to stay true, and now something asks. */
  ['text/x-component', false, '0:{"a":1}\n', 'text/x-component',
   '§7 reaches step 8 for a supplied type no rule touches, so a Flight stream is what its server said it is'],
  /* A HEADER PRESENT WITH AN EMPTY VALUE IS NOT AN ABSENT HEADER, even though §5.1 reaches the same undefined
     supplied type for both. The record can say each, which is the point; if it could not, the fact would be
     lost at the boundary rather than at the step that stops caring about it. */
  ['', false, GIF, 'image/gif',
   "an empty Content-Type value parses as no MIME type, which is §5.1's undefined supplied type"],
  /* A RESPONSE WITH NO BODY AT ALL is §5.2's "the end of the resource is reached" and not an error. */
  [null, false, '', 'text/plain',
   '§7.1 over an empty resource header finds no binary data byte and answers text/plain'],
];

let failed = 0;
for (const [ct, noSniff, body, want, why] of CASES) {
  const got = sniff(ct, noSniff, body);
  const shown = ct === null ? '(no Content-Type header)' : JSON.stringify(ct);
  if (got === want) {
    console.log(`  ok   ${shown}${noSniff ? ' +nosniff' : ''} -> ${got}`);
  } else {
    failed++;
    console.error(`  FAIL ${shown}${noSniff ? ' +nosniff' : ''} -> ${got}   expected ${want}\n       ${why}`);
  }
}

/* ── the boundary, measured ────────────────────────────────────────────────────────────────────────────── */

const child = spawnSync(process.execPath, [fileURLToPath(import.meta.url), '--as-renderer'],
                        { encoding: 'utf8' });
const said = (child.stderr || '') + (child.stdout || '');
/* THE ABORT AND WHAT IT SAID, both. A non-zero exit on its own is satisfied by any crash — a missing artifact,
   a bad ccall — so the `@E` line is what says the abort was THIS boundary and not an accident on the way to it. */
const aborted = child.status !== 0;
const named = said.includes('@E {');        /* check.h's always-fatal tag, emitted before the abort */
const zoneOnly = said.includes('zone-only');
if (aborted && named && zoneOnly) {
  console.log('  ok   a renderer asking a zone-only operation aborted the browser process');
} else {
  failed++;
  console.error('  FAIL a renderer asked `mime.sniff` and the browser process ' +
                (aborted ? 'aborted without naming the boundary' : 'ANSWERED IT') +
                ` (exit ${child.status})\n${said.split('\n').slice(-6).join('\n')}`);
}

if (failed) {
  console.error(`[browser-process] FAILED: ${failed} case(s) — each names the §7 step it disagrees with, and a ` +
                'disagreement is the standard being read wrong here rather than a case being wrong there');
  process.exit(1);
}
console.log(`[browser-process] OK — one browser-process instance, ${CASES.length} §7 answers, ` +
            'and a renderer refused');
